### PR TITLE
Mark’s Inject Challenge Submission

### PR DESCRIPTION
**Pull request by Gwasanaethau**
Mark Pokorny’s (Gwasanaethau’s) submission for the Inject Challenge.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,25 +1,3 @@
...
-* Name your method differently (that is, not inject() or subclass Array) because rspec uses inject() internally, so the tests will fail with weird messages unless your implementation of inject is perfect.
-* If you would like a challenge, rewrite inject using two approaches: using iterators and using recursion but it’s not required to pass.
-* Submit pull request with your solution
-
-
-[![Build Status](https://travis-ci.org/makersacademy/inject-challenge.svg?branch=master)](https://travis-ci.org/makersacademy/inject-challenge)
-[![Coverage Status](https://coveralls.io/repos/makersacademy/inject-challenge/badge.png)](https://coveralls.io/r/makersacademy/inject-challenge)
+The purpose of the Inject Challenge is to test the knowledge of and skills in TDD gained over the previous few weeks, and using knowledge of blocks, procs and lambdas gained during the precourse to replicate the ‘inject’ method in the ‘Array’ class. The challenge is outlined in the [Instructions](./Instructions.md) and is to be completed over the weekend, and the progress submitted by 09:00 on Monday 30th March.
```
Comment (sjmog) [inline README.md:6]: Great readme, once again :)

File: spec/array_spec.rb [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    expect([1, 2, 3, 4, 5].custom_inject(100, :-)).to eq 85
+  end
+
+  it 'uses inject with initial value and symbol to divide elements…' do
+    expect([30, 5, 2].inject(600, :/)).to eq 2
+  end
+  it 'uses custom_inject with initial value to divide elements…' do
+    expect([30, 5, 2].custom_inject(600, :/)).to eq 2
+  end
+end
```
Comment (sjmog) [inline spec/array_spec.rb:139]: A very thorough spec :+1:

File: lib/array.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+class Array
+  def custom_inject(*args, &block)
+    operation, injector = retrieve_args(*args)
+    injector ||= shift # i.e: self.shift (RuboCop is not being friendly!)
```
Comment (sjmog) [inline lib/array.rb:4]: What do you mean here? Since we are in the class `self` is implicit

File: lib/array.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    operation_select(operation, injector, block)
+  end
+
+  private
+
+  def retrieve_args(*args)
+    symbol = args.select { |arg| arg.is_a?(Symbol) }.first
+    initial_value = args.select { |arg| arg.is_a?(Numeric) }.first
+    [symbol, initial_value]
+  end
```
Comment (sjmog) [inline lib/array.rb:14]: A beautiful method, great OOP

File: lib/array.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
+class Array
+  def custom_inject(*args, &block)
+    operation, injector = retrieve_args(*args)
+    injector ||= shift # i.e: self.shift (RuboCop is not being friendly!)
+    operation_select(operation, injector, block)
+  end
+
+  private
```
Comment (sjmog) [inline lib/array.rb:8]: Excellent use of `private`

File: lib/array.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    initial_value = args.select { |arg| arg.is_a?(Numeric) }.first
+    [symbol, initial_value]
+  end
+
+  def operation_select(operation, injector, block)
+    case operation
+    when :+ then each { |num| injector += num }
+    when :- then each { |num| injector -= num }
+    when :* then each { |num| injector *= num }
+    when :/ then each { |num| injector /= num }
```
Comment (sjmog) [inline lib/array.rb:21]: Can you think of a more DRY way to do this? How about something that uses `.send`?

**Comment by sjmog** [comment]
A pretty epic solution, Mark - and that's one hell of a spec file :+1: 

I love how you're breaking things down so methods are following single responsibility :+1: 

Great solution!

:checkered_flag:

File: lib/array.rb (hunk 3 above)
Comment (Gwasanaethau) [inline lib/array.rb:4]: Very true, but it seemed weird to me to see ‘shift’ without something in front of it (bad Java perspective, perhaps?).

File: lib/array.rb (hunk 6 above)
Comment (Gwasanaethau) [inline lib/array.rb:21]: I had a quick look at ‘.send’ as I was writing this but started running low on time, so went with the method I knew that worked. I might refactor if I get time. Thanks for the suggestion! :)

File: lib/array.rb (hunk 3 above)
Comment (sjmog) [inline lib/array.rb:4]: Yep! You can and should use implicit `self` wherever possible.

File: lib/array.rb (hunk 6 above)
Comment (sjmog) [inline lib/array.rb:21]: No problem!

**Pull request closed without merging**